### DISCUSSION TITLE
How to use postUpgradeTasks

### DISCUSSION THREAD

**Discussion by 1oglop1** [renovatebot/renovate#21247] [Request Help] [answered]
### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us what version of Renovate you run.

npm 34.158.1 or gitlab latest

### If you're self-hosting Renovate, select which platform you are using.

gitlab.com

### Was this something which used to work for you, and then stopped?

I never saw this working

### Wanted end result.

I have the following structure in my target repository

```
.tool-versions
.gitlab-ci.yaml
```

content of `.tool-versions`
```
pnpm 7.28.0
```

content of `.gitlab-ci.yaml`
```yaml
variables:
  PNPM_VERSION: 7.28.0
stages:
  - test 
    
my-job:
  stage: test
  script:
    - npm i -g pnpm@${PNPM_VERSION}
    - pnpm --version
```

I'd like to make a postUpgrade command to update `.gitlab-ci.yaml` file together with `.tool-versions` file because they need to be kept in sync.

However, I am not able to get even basic postUpgrade command to run.

### What you tried so far.

`renovate.json5`
```
{
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "extends": [
    "config:base"
  ],
  enabledManagers: ["asdf"],
//  "packageRules": [
//      {
//        "matchManagers": ["asdf"],
//        "matchPackageNames": ["pnpm"],
//         // I also tried to put section here postUpgradeTasks but nothing has changed
//      } 
//    ],
  "postUpgradeTasks": {
            "commands": ["touch $(date +%Y%m%d_%H%M%S)"],
            "executionMode": "branch", // I tried "update" as well
            "fileFilters": ["**/**"]
          }

}

```

`config.js`
```js
module.exports = {
  autodiscover: true,
  "autodiscoverFilter": [
    "myorg/myRepo"
  ],
  "onboarding": false,
  allowedPostUpgradeCommands: ["^touch"],
  allowPostUpgradeCommandTemplating: true,
};
```

### Relevant debug logs

[Logs](https://gist.github.com/1oglop1/02e9fa3987dd144a16532092787c9b3e)

**Comment by rarkins** [comment]
I recommend that you instead use a regex manager for the gitlab ci file. If you have the same `depName` for asdf's pnpm as well as your regex manager then they will be upgraded together

**Reply by 1oglop1** [reply]
Thanks, I thought about that. Let me look at the docs and see if I can make it work.

Is there a way to test all this without making commits and running cli?

**Reply by 1oglop1** [reply]
In the end, I got it working!

Used the example https://docs.renovatebot.com/modules/manager/regex/#advanced-capture

And modified regex from https://docs.renovatebot.com/presets-regexManagers/#regexmanagersgithubactionsversions
 to match `.gitlab-ci.yml` file.


```
   regexManagers: [
    {
      "fileMatch": [
        "^\.gitlab-ci\.ya?ml$"
      ],
      "matchStrings": [
        "# renovate: datasource=(?<datasource>[a-z-]+?) depName=(?<depName>[^\\s]+?)(?: (?:lookupName|packageName)=(?<packageName>[^\\s]+?))?(?: versioning=(?<versioning>[a-z-0-9]+?))?\\s+[A-Za-z0-9_]+?_VERSION\\s*:\\s*[\"']?(?<currentValue>.+?)[\"']?\\s"
      ]
    }]
    ```


And updated my gitlabCI file with renovate comment:
```
variables:
  # renovate: datasource=npm depName=pnpm versioning=semver
  PNPM_VERSION: 7.28.0
  ```

**Comment by 1oglop1** [comment, marked as answer]
@rarkins Thank you! I managed to get almost everything. 
Now I need to run `pnpm i`. 
Because I have `.pnpmfile.cjs` hook file to modify dependencies of sub-packages and when renovate makes a PR it updates the lockfile incorrectly.  At the moment I am not sure if it is because pnpm runs in CI mode or what is wrong. 

- I fixed this by `allowScripts:true` in `config.js` and `ignoreScripts: false` in `renovate.json`.

--- 


Another issue is that grouping does not work as expected. 
I do have this to group `asdf` and `npm` under one group - this works as expected, except that `asdf` keep creating branch `renovate/pulumi-3.x` so in the end there are 2 branches `renovate/pulumi-3.x` and `renovate/pulumi`
```
  packageRules: [
    {
      groupName: "pulumi",
      matchDepNames: [
        "@pulumi/pulumi", // npm
        "pulumi/pulumi" // asdf
      ]
    }
  ],
  ```
I tried to remove `groupName` and see if it helps

EDIT:
In the end, I managed to unify the update after carefully reading the logs I found a message `Found sourceUrl with multiple branches that should probably be combined into a group` and changed the `matchDepNames` -> `matchSourceUrl: ["https://github.com/pulumi/pulumi"]` after that, PR finally grouped the dependency together.
